mlir/lib/Catalyst/Transforms/BufferizableOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,289 @@
+    }
+
+    bool bufferizesToMemoryWrite(Operation *op, OpOperand &opOperand,
+                                 const bufferization::AnalysisState &state) const
+    {
+        return false;
+    }
+
+    bufferization::AliasingValueList
+    getAliasingValues(Operation *op, OpOperand &opOperand,
+                      const bufferization::AnalysisState &state) const
+    {
+        return {};
+    }
+
+    LogicalResult bufferize(Operation *op, RewriterBase &rewriter,
+                            const bufferization::BufferizationOptions &options) const
+    {
+        auto customCallOp = cast<CustomCallOp>(op);
+
+        // Add bufferized arguments
+        SmallVector<Value> bufferArgs;
+        ValueRange operands = customCallOp.getOperands();
+        for (Value operand : operands) {
+            FailureOr<Value> opBuffer = getBuffer(rewriter, operand, options);
+            if (failed(opBuffer))
+                bufferArgs.push_back(operand);
+            else
+                bufferArgs.push_back(*opBuffer);
+        }
+
+        // Add bufferized return values to the arguments
+        ValueRange results = customCallOp.getResults();
+        for (Value result : results) {
+            Type resultType = result.getType();
+            RankedTensorType tensorType = dyn_cast<RankedTensorType>(resultType);
+            if (!tensorType) {
+                bufferArgs.push_back(result);
+                continue;
+            }
+            auto options = bufferization::BufferizationOptions();
+            FailureOr<Value> tensorAlloc = bufferization::allocateTensorForShapedValue(
+                rewriter, op->getLoc(), result, options, false);
+            MemRefType memrefType =
+                MemRefType::get(tensorType.getShape(), tensorType.getElementType());
+            auto newBuffer =
+                rewriter.create<bufferization::ToMemrefOp>(op->getLoc(), memrefType, *tensorAlloc);
+            bufferArgs.push_back(newBuffer);
+        }
+
+        // Add the initial number of arguments
+        int32_t numArguments = static_cast<int32_t>(customCallOp.getNumOperands());
+        DenseI32ArrayAttr numArgumentsDenseAttr = rewriter.getDenseI32ArrayAttr({numArguments});
+
+        // Create an updated custom call operation
+        rewriter.create<CustomCallOp>(op->getLoc(), TypeRange{}, bufferArgs,
+                                      customCallOp.getCallTargetName(), numArgumentsDenseAttr);
+        size_t startIndex = bufferArgs.size() - customCallOp.getNumResults();
+        SmallVector<Value> bufferResults(bufferArgs.begin() + startIndex, bufferArgs.end());
+        bufferization::replaceOpWithBufferizedValues(rewriter, op, bufferResults);
+
+        return success();
+    }
+};
+
+struct CallbackOpInterface
+    : public bufferization::OpWithUnstructuredControlFlowBufferizableOpInterfaceExternalModel<
+          CallbackOpInterface, CallbackOp> {
+    static bool supportsUnstructuredControlFlow() { return true; }
```

Comment:
We never use unstructured control flow in callback op. I've set this to false and it works. Can you expand on why this is needed?

Reply:
I was following the `FuncOpInterface`. I will set to as false.

Reply:
Can you show also why exactly you needed to implement this interface? I am curious. Thanks!

Reply:
I tried to make the interface similar to FuncOp's. https://github.com/llvm/llvm-project/blob/f59b0c76030aff268b78d475e219708d06b982b5/mlir/lib/Dialect/Bufferization/Transforms/FuncBufferizableOpInterfaceImpl.cpp#L325-L351

According to this LLVM commit https://github.com/llvm/llvm-project/commit/6ecebb496cc6960e100a05375ab7f64e831dd933,
we might only need it if `cf.if` appears in `func.func`. I remember last time I tested it and function like ops did not work without this interface.
I will try using the original interface `BufferizableOpInterface::ExternalModel` again to see what is going on.

Reply:
It seems we do not need this `OpWithUnstructuredControlFlowBufferizableOpInterfaceExternalModel` for CallbackOp. I will remove it.

Reply:
Yes, this data layout issue should be removed by identity-layout-map flag. Not sure why this error comes back even though the flag is already set.